src/util/coding-jams/managePolls.ts

```diff
@@ -0,0 +1,290 @@
+import { getFromEnmap } from "../misc/enmap";
+import { discordRelativeTimestamp, discordTimestamp } from "../misc/time";
+
+export async function newPoll(
+	interaction: CommandInteraction,
+	name: string,
+	numProposals: number,
+	numVotes: number,
+	selectionType: string,
+	start: DateTime,
+	end: DateTime
+) {
+	if (pollDb.findKey((poll) => poll.title === name)) {
+		interaction.reply({
+			content: `A poll with the name "${name}" already exists...`,
+			ephemeral: true,
+		});
+		return;
+	}
+
+	if (proposalDb.size < 2) {
+		interaction.reply({
+			content: "There simply aren't enough proposals to create a poll...",
+			ephemeral: true,
+		});
+		return;
+	}
+
+	// check numVotes and numProposals
+	if (numVotes > numProposals) numVotes = numProposals;
+	if (proposalDb.size < numProposals) {
+		numProposals = proposalDb.size;
+		numVotes = proposalDb.size;
+	}
+
+	const sorted = sortBySelectionType(selectionType);
+
+	const poll: Poll = {
+		title: name,
+		numProposals: numProposals,
+		numVotes: numVotes,
+		start: start,
+		end: end,
+		selectionType: selectionType,
+		exclude: [],
+		include: [],
+		votingPrompt: null,
+		proposals: sorted.slice(0, numProposals).map((e) => e.value),
+		votes: new Map<string, string[]>(),
+	};
+
+	const id = pollDb.autonum;
+	pollDb.set(id, poll);
+
+	const events: PollEvent[] = [
+		{ pollID: id, type: "before", promptID: null, date: start.minus({ days: 1 }) },
+		{ pollID: id, type: "open", promptID: null, date: start },
+	];
+
+	events.forEach((e) => pollEventsDb.set(pollEventsDb.autonum, e));
+
+	interaction.reply({
+		embeds: [pollEmbed(poll, id, "(new)")],
+		components: pollSelectMenus(poll, id, sorted, sorted),
+		ephemeral: true,
+	});
+}
+
+export async function editPoll(interaction: CommandInteraction, name: string, end: DateTime) {
+	const pollKey = pollDb.findKey((poll) => poll.title === name);
+
+	if (!pollKey) {
+		interaction.reply({ content: "There is no poll with that name...", ephemeral: true });
+		return;
+	}
+
+	const poll = pollDb.get(pollKey)!;
+
+	if (poll.end < DateTime.now()) {
+		await interaction.reply({ content: "You can't extend polls that have already ended!", ephemeral: true });
+		return;
+	}
+
+	pollDb.set(pollKey, { ...poll, end: end });
+
+	const diff = poll.end.diff(end);
+	const inFuture = poll.end < end;
+
+	for (const key of pollEventsDb.keyArray()) {
+		const poll = pollEventsDb.get(key)!;
+		if (!(poll.pollID === pollKey && poll.type === "close")) continue;
+		pollEventsDb.update(key, (event) => {
+			const old = event.date;
+			const newDate = inFuture ? old.plus(diff) : old.minus(diff);
+			return { ...event, date: newDate };
+		});
+	}
+
+	await interaction.reply({
+		embeds: [pollEmbed(poll, pollKey, "(edit)")],
+		components: pollSelectMenus(
+			poll,
+			pollKey,
+			sortBySelectionType(poll.selectionType).filter((e) => !poll.exclude.includes(e.value)),
+			sortBySelectionType(poll.selectionType).filter((e) => !poll.include.includes(e.value))
```

Comment:
Why run the same function + a filter twice?

Reply:
the filter is different but yeah can do the sorting once